### ISSUE TITLE
[Umbrella] Bill of Materials

### ISSUE THREAD

**Issue by puerco** [kubernetes/release#1837]
We intend to ensure the quality and integrity of the artifacts produced on each release cut by adding a Bill of Materials (BOM). The BOM will be published in [SPDX](https://spdx.dev/) and will include integrity and licensing information for the artifacts we produce. Work on this area will lead to close other outstanding issues (linked here).

Following our road-mapping session, this umbrella issue will track the development to create the BOM.

### ~Make krel aware of binary artifacts expected from the release process:~
**Note:** These items are postponed as we delayed the supported platforms effort to 1.23+
- [x] ~Read the data from the proposed (https://github.com/kubernetes/release/issues/1836) machine-readable platform map~
- [x] ~Bootstrap the Release Process [state object](https://github.com/kubernetes/release/blob/6b7d184104018f8968176c0b0fedae77266ce098/pkg/anago/anago.go#L129-L142) with files expected as output, crossing the platform data and options specified in the run.~

### Verify/process binary artifacts as the release process advances from stage to stage
- [x] Verify binaries to ensure the platform is correct (https://github.com/kubernetes/release/issues/1521 #2160 ) 
This step involves:
  - [x] Creating a new API in our tools to check the binaries (@puerco #1856)
  - [x] Integrating it into the release process to check compiled artifacts (#2160)
- [x] Verify tagging in binaries (issue https://github.com/kubernetes/release/issues/1898) (#2160)
We need to ensure that binaries are correctly tagged with the corresponding semver tag and commit sha

### Write SPDX manifest(s). Output should include data about:
- [x] Naked binaries (#2095)
- [x] Images (https://github.com/kubernetes/release/issues/1384) (#2095)
- [x] Third-party licences (https://github.com/kubernetes/kubernetes/issues/94976) . 
To accomplish this one we will need to:
  - [x] Write or use a license parsing tool to obtain licenses of external modules. (https://github.com/kubernetes/release/pull/1874 @puerco) 
  - [x] ~Integrate the scanner into the release process scanning the [vendor/ directory](https://github.com/kubernetes/kubernetes/tree/master/LICENSES) in k/k to get the licenses that will be added to the BOM~ Scan all dependencies licensing information and include them in the SBOM (#2095)

### Publish the SPDX manifests with the other release artifacts:
- [x] Generate the BOM in one or more formats supported by SPDX: [tag](https://spdx.dev/resources/use/#documents), RDF, etc. (#2095)
- [x] Write them to a subdirectory in the release bucket (#2095)
- [x] ~Upload manifests as assets in the GitHub release page~ **Note:** In later discussions we chose to publish the documents only via https for now and not relay on the GH release page. 
- [x] Publish the SBOM documents via sbom.k8s.io (https://github.com/kubernetes/k8s.io/pull/2447)

### Make our tools available community-wide
- [x] Build a general-purpose tool to generate SPDX compliant SBOMs (#2066)
- [x] Support defining complex SBOM compositions from a yaml file (#2096)
- [x] Add bom utility documentation (https://github.com/kubernetes/release/pull/2109)
- [x] Write an in-depth guide to generating SPDX bill of materials (https://github.com/kubernetes/release/pull/2109)

/cc @hasheddan @xmudrii @markyjackson-taulia

**Comment by nishakm** [comment]
cc @rnjudge and @kestewart for Tern

**Comment by tsteenbe** [comment]
@puerco If you need a tool to generate SPDX for Go and other languages, have a look at [OSS Review Toolkit](https://github.com/oss-review-toolkit/ort). I am both ORT and SPDX maintainer, happy to answer any questions you may have.

**Comment by puerco** [comment]
Thank you @tsteenbe I certainly will take a look at it :)

**Comment by justaugustus** [comment]
Connecting some threads based on an OSS supply chain convo I had earlier in the week with folks across VMware, Red Hat, and Google (@ncdc, @dlorenc, @wattsteve, @lukehinds, @kimsterv, @jonjohnsonjr to name a few):

- https://github.com/opencontainers/image-spec/blob/master/spec.md
- https://github.com/opencontainers/distribution-spec/blob/master/spec.md
- https://github.com/opencontainers/artifacts/pull/29
- https://github.com/cncf/sig-security/issues/467#issuecomment-740573511
- https://rekor.dev/
- https://rekor.dev/docs/plugable_types/

@saschagrunert @hasheddan @puerco -- Let's please leverage Rekor here as we build this out.

Of course, there's more to consider and supply chain concerns will be core to everything we do in RelEng, but the bill of materials is a great starting point.

cc: @kubernetes/sig-security-leads

**Comment by lukehinds** [comment]
Rekor could in particular be beneficial for 'Publish the SPDX manifests with the other release artifacts' - this would provide two key benefits off the top of my head.

1. immutability, making the BOMs tamper resistant.
2. Public and assured transparency, e.g. more eyes to monitor the BOMs for anything untoward 

I will read up on SPDX (tagging @bobcallaway for types expertise). Do you know if there will be some sort of signature at BOM generation time?

**Comment by puerco** [comment]
@lukehinds the first iteration of the SPDX BOM will not be signed as we have yet to set a key management mechanism before we start signing the manifests and other release artifacts. We hope to have it resolved soon(ish) as this is the main issue [blocking us from generating and hosting debs and rpms in community infra].(https://github.com/kubernetes/release/issues/281#issuecomment-774578036).

**Comment by saschagrunert** [comment]
@puerco is this one done?

**Comment by saschagrunert** [comment]
@puerco can we close this one as done?

**Comment by puerco** [comment]
The SBOM project is done. We are now producing SBOMs with all k8s releases:

https://sbom.k8s.io/v1.23.0/release
https://sbom.k8s.io/v1.23.0/source

And we have spun off the took into its own repository: https://sigs.k8s.io/bom

Thanks for all your help and support, everybody! :partying_face: 

/close
